dask/layers.py

```diff
@@ -847,3 +867,140 @@ def _construct_graph(self, deserializing=False):
             dsk[(self.name, i)] = (concat_func, _concat_list)
 
         return dsk
+
+
+class MaterializedIODeps(BlockwiseIODeps):
+    """Partially-materialized IO-Dep Wrapper"""
+
+    def __init__(self, parts):
+        self.parts = parts
+
+    def __getitem__(self, idx: tuple):
+        return self.parts[idx]
+
+    @classmethod
+    def __dask_distributed_pack__(cls, cls_path: str, parts: dict):
+        import pickle
+
+        packed_parts = {}
+        for k, v in parts.items():
+            packed_parts[k] = pickle.dumps(v)
+
+        return (cls_path, packed_parts)
```

Comment:
Typical IO layers will use a materialized `dict` (currently called `parts` within all the DataFrame-IO Layer defininitions) to define the elements of `Blockwise.io_deps`.  There is usually no reason to use the `BlockwiseIODeps` approach, because the client already needs to generate most of the required IO dependencies while it is calculating the mapping between byte-ranges and output partitions (we dont even know how many partitions there will be until this is done).  The only time the `BlockwiseIODeps` approach makes sense is when we know that some elements of the materialized `dict` will likely contain non-msgpack-serializable objects.  In these cases, we use this `MaterializedIODeps` class to make sure the each element of `parts` is serialized with `pickle`.

The `MaterializedIODeps` approach is always used by read_csv in this PR, and it is also used by read_parquet when the user specifies `large_graph_objects=True`.  We may be able to avoid pickling in`read_csv`, but we are current relying on it to serialize the `delayed` text block argument to the final IO function.

Reply:
Update - `MaterializedIODeps` is now using `distributed.protocol.serialize` instead of pickle.  This is effectively the same thing, but establishes a convention that will allow special/complex types to be serialized properly in the future.